File: 978.js
/**
 * @param {number[]} A
 * @return {number}
 */
var maxTurbulenceSize = function(A) {
  if (A.length === 1 || A.length === 2) {
    return A.length;
  }

  var max = 2;
  for (var i = 0; i < A.length - 1; i++) {
    var cur = 2;
    var diff = A[i + 1] - A[i];
    while (i < A.length - 1) {
      var newDiff = A[i + 2] - A[i + 1];
      if (diff * newDiff >= 0 || i + 2 === A.length) {
        max = Math.max(max, cur);
        break;
      }
      diff = newDiff;
      cur++;
      i++;
    }
  }
  return max;
};

console.log(maxTurbulenceSize([4, -1, 1, -1, -1, 1, 0, 1, 0, 1]));

File: 978.java
class Solution978 {
  public static void main(String args[]) {
    Solution978 s = new Solution978();
    s.maxTurbulenceSize(new int[] { 2, 0, 2, 4, 2, 5, 0, 1, 2, 3 });
  }

  public int maxTurbulenceSize(int[] arr) {
    if (arr.length == 0 || arr.length == 1)
      return arr.length;
    int max = 1;
    // 0 same, 1 got bigger, -1 got smaller.
    int status = 0;
    int prev_val = arr[0];
    int cur_len = 1;
    for (int i = 1; i < arr.length; i++) {
      if (prev_val == arr[i]) {
        status = 0;
        max = Math.max(max, cur_len);
        cur_len = 1;
      } else if (prev_val < arr[i]) {
        if (status == -1) {
          cur_len++;
        } else {
          max = Math.max(max, cur_len);
          cur_len = 2;
        }
        status = 1;
      } else {
        if (status == 1) {
          cur_len++;
        } else {
          max = Math.max(max, cur_len);
          cur_len = 2;
        }
        status = -1;
      }
      prev_val = arr[i];
    }
    return Math.max(max, cur_len);
  }
}